csharp/src/Ice/Endpoint.cs

```diff
@@ -131,5 +133,124 @@ public abstract class Endpoint : IEquatable<Endpoint>
+                string option = args[n];
+                if (option.Length < 2 || option[0] != '-')
+                {
+                    throw new FormatException($"invalid option `{option}' in endpoint `{endpointString}'");
+                }
+
+                // Extract the argument given to the current option, if any
+                string? argument = null;
+                if (n + 1 < args.Length && args[n + 1][0] != '-')
+                {
+                    argument = args[++n];
+                }
+
+                try
+                {
+                    options.Add(option, argument);
+                }
+                catch (ArgumentException)
+                {
+                    throw new FormatException($"duplicate option `{option}' in endpoint `{endpointString}'");
+                }
+            }
+
+            if (communicator.FindEndpointFactory(transport) is IEndpointFactory factory)
+            {
+                Endpoint endpoint = factory.Create(endpointString, options, oaEndpoint);
+                if (options.Count > 0)
+                {
+                    throw new FormatException(
+                        $"unrecognized option(s) `{ToString(options)}' in endpoint `{endpointString}'");
+                }
+                return endpoint;
+            }
+
+            // If the stringified endpoint is opaque, create an unknown endpoint, then see whether the type matches one
+            // of the known endpoints.
+            if (!oaEndpoint && transport == "opaque")
+            {
+                var opaqueEndpoint = new OpaqueEndpoint(endpointString, options);
+                if (options.Count > 0)
+                {
+                    throw new FormatException(
+                        $"unrecognized option(s) `{ToString(options)}' in endpoint `{endpointString}'");
+                }
+
+                if (opaqueEndpoint.Encoding.IsSupported &&
+                    communicator.FindEndpointFactory(opaqueEndpoint.Type) != null)
+                {
+                    // We may be able to unmarshal this endpoint, so we first marshal it into a byte buffer and then
+                    // unmarshal it from this buffer.
+                    var bufferList = new List<ArraySegment<byte>>
+                    {
+                        // 8 = size of short + size of encapsulation header
+                        new byte[8 + opaqueEndpoint.Bytes.Length]
+                    };
+
+                    var ostr = new OutputStream(Ice1Definitions.Encoding, bufferList);
+                    ostr.WriteEndpoint(opaqueEndpoint);
+                    OutputStream.Position tail = ostr.Save();
+                    Debug.Assert(bufferList.Count == 1);
+                    Debug.Assert(tail.Segment == 0 && tail.Offset == 8 + opaqueEndpoint.Bytes.Length);
+
+                    return new InputStream(communicator, bufferList[0]).ReadEndpoint();
+                }
+                else
```

Comment:
Ok, I guess it's because we don't want to return the opaque endpoint for an object adapter... so perhaps the `if(!oaEndpoint)` should be on this `else`.